### PR TITLE
feat: length reward

### PR DESCRIPTION
**Pull request by justusmattern27**

### CODE REVIEW

File: src/zeroband/inference.py [hunk 1]
```diff
@@ -75,28 +75,20 @@ class Config(BaseConfig):
 
     ckpt_start_path: str | None = None
 
+    length_reward_min: int | None = None
+    length_reward_max: int | None = None
+    length_reward_coeff: float | None = None
```
Comment (samsja) [inline src/zeroband/inference.py]:
```suggestion
    len_reward_min: int | None = None
    len_reward_max: int | None = None
    len_reward_coeff: float | None = None
```

preventing future typo :p

Comment (samsja) [reply src/zeroband/inference.py]: lets rather do 

```pyton

class LenRewardConfig(Config):
    reward_min : int  = # add default value here
    reward_max : int  = # add default value here
    reward_coef : int  = # add default value here

class Config(BaseConfig):
    ...
    
    len_reward : LenRewardConfig | None = None

File: src/zeroband/train.py [hunk 2]
```diff
@@ -196,11 +195,22 @@ def train(config: Config):
         num = 1 if isinstance(config.train.ac_ckpt, bool) else config.train.ac_ckpt
         apply_ac_ckpt(model, num)
 
-    apply_fsdp(model, config.train.reshard_after_forward)
+    model = apply_fsdp(model)
```
Comment (samsja) [inline src/zeroband/train.py]: this should not change here

File: src/zeroband/train.py [hunk 3]
```diff
@@ -196,11 +195,22 @@ def train(config: Config):
...
 
-    optimizer = torch.optim.AdamW(params=model.parameters(),lr=config.optim.optim.lr,weight_decay=config.optim.optim.weight_decay,betas=(config.optim.optim.betas1, config.optim.optim.betas2))  # fmt: skip
+    optimizer = torch.optim.AdamW(
+        params=model.parameters(),
+        lr=config.optim.optim.lr,
+        weight_decay=config.optim.optim.weight_decay,
+        betas=(config.optim.optim.betas1, config.optim.optim.betas2),
+    )
```
Comment (samsja) [inline src/zeroband/train.py:205]: should not change neither

File: src/zeroband/train.py [hunk 4]
```diff
@@ -240,18 +250,19 @@ def train(config: Config):
 
                 for rollout_step in range(config.optim.step_per_rollout):
                     for grad_acc_step in range(gradient_accumulation_steps):
-                        batch = next(train_dataloader_iterator)
-                        input_ids = batch["input_ids"].to("cuda")
+                        with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
+                            batch = next(train_dataloader_iterator)
```
Comment (samsja) [inline src/zeroband/train.py:254]: we don't need autocast anymore

File: src/zeroband/train.py [hunk 5]
```diff
@@ -270,33 +281,38 @@ def train(config: Config):
...
                 batch = next(logprobs_aware_iterator)
                 input_ids = batch["input_ids"].to("cuda")
                 loss_mask = batch["loss_mask"]
 
-                rewards = batch["rewards"][loss_mask.bool()]
-                rewards_sum += rewards.sum()
-                rewards_token_count += rewards.numel()
+                rewards_this = batch["rewards"][loss_mask.bool()]
+                rewards_sum += rewards_this.sum()
+                rewards_token_count += rewards_this.numel()
```
Comment (samsja) [inline src/zeroband/train.py:303]: what does this stand for here ?

File: src/zeroband/train.py [hunk 6]
```diff
@@ -311,13 +327,41 @@ def train(config: Config):
...
+                )
 
-                sample_reward_batch += batch["rewards"][:, 0].sum() / batch["rewards"].shape[0] / gradient_accumulation_steps
+                seq_lens_batch += batch["seq_lens"].float().mean() / gradient_accumulation_steps
+                clip_seq_lens += (
+                    (batch["seq_lens"] >= config.data.seq_length).sum() / batch["seq_lens"].shape[0] / gradient_accumulation_steps
+                )
 
-                del batch, logits, input_ids, advantages, loss_mask, original_logprobs
+                with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
```
Comment (samsja) [inline src/zeroband/train.py:339]: need to remove autocast

File: src/zeroband/training/data.py [hunk 7]
```diff
@@ -179,68 +186,110 @@ def __iter__(self):
 
             dataset = _get_dataset_from_files_step(self._step_count, self._path, self._timeout, self._batch_size)
 
-            # we are NOT splitting the files across datalaoder workers and rank like we did for intellect 1
-            # This is because we cannot assume that the files would have the same number of samples each.
-            # What we rather do here is that all the workers go over all the files and only yield some of them
-            # this is unoptimal because they all load more data that they should, but since the data is already tokenized it should not be a big deal
-
-            # Set up a scanner with just the required columns
-            required_columns = ["input_tokens", "output_tokens", "advantages", "rewards", "input_logprobs", "output_logprobs"]
```
Comment (samsja) [inline src/zeroband/training/data.py:188]: lets keep the comment ?

File: src/zeroband/training/data.py [hunk 8]
```diff
@@ -179,68 +186,110 @@ def __iter__(self):
...
+                        output_logprobs,
+                        advantages,
+                        rewards,
+                        task_rewards,
+                        length_penalties,
+                        target_lengths,
                     ):
                         counter += 1
-                        if not _should_skip_index(
+                        if _should_skip_index(
```
Comment (samsja) [inline src/zeroband/training/data.py:238]: what is the reason the remove the not here ?

**Comment by samsja** [comment]
closing as it was rebase and merge here https://github.com/PrimeIntellect-ai/prime-rl/pull/133
